src/scanner.cc

```diff
@@ -0,0 +1,319 @@
+#include <tree_sitter/parser.h>
+#include <stdio.h>
+
+enum TokenType : TSSymbol {
+  SIMPLE_STRING,
+  SIMPLE_SYMBOL,
+  SIMPLE_SUBSHELL,
+  SIMPLE_REGEX,
+  SIMPLE_WORD_LIST,
+  STRING_BEGINNING,
+  SYMBOL_BEGINNING,
+  SUBSHELL_BEGINNING,
+  REGEX_BEGINNING,
+  WORD_LIST_BEGINNING,
+  STRING_MIDDLE,
+  STRING_END
+};
+
+enum LiteralType {
+  STRING,
+  SYMBOL,
+  SUBSHELL,
+  REGEX,
+  WORD_LIST
+};
+
+TokenType BEGINNING_TOKEN_TYPES[] = {
+  STRING_BEGINNING,
+  SYMBOL_BEGINNING,
+  SUBSHELL_BEGINNING,
+  REGEX_BEGINNING,
+  WORD_LIST_BEGINNING,
+};
+
+TokenType SIMPLE_TOKEN_TYPES[] = {
+  SIMPLE_STRING,
+  SIMPLE_SYMBOL,
+  SIMPLE_SUBSHELL,
+  SIMPLE_REGEX,
+  SIMPLE_WORD_LIST,
+};
+
+struct Scanner {
+  Scanner() :
+    literal_type{STRING}, open_delimiter{0}, close_delimiter{0}, depth{0} {}
+
+  void advance(TSLexer *lexer, bool is_whitespace = false) {
+    lexer->advance(lexer, is_whitespace);
+  }
+
+  bool scan_whitespace(TSLexer *lexer) {
+    while (lexer->lookahead == ' ' ||
+           lexer->lookahead == '\t' ||
+           lexer->lookahead == '\n') {
```

Comment:
I think we might need to add scanning for `\` followed by a newline in here so that we can parse things like;

``` ruby
a = "this string \
is on multiple lines"
```